File: haskell/2020/Day2.hs
module Day2 where

import Data.Algebra.Boolean qualified as B
import Data.List.Split

parsePassword :: [String] -> (Int, Int, Char, String)
parsePassword [r, c, pw] = (lo, hi, head c, pw)
  where
    [lo, hi] = map read $ splitOn "-" r

validate :: (Int, Int, Char, String) -> Bool
validate (lo, hi, c, pw) = cs >= lo && cs <= hi
  where
    cs = length . filter (== c) $ pw

validate2 :: (Int, Int, Char, String) -> Bool
validate2 (lo, hi, c, pw) = (p1 == c) `B.xor` (p2 == c)
  where
    p1 = pw !! (lo - 1)
    p2 = pw !! (hi - 1)

main' :: IO ()
main' = do
  content <- map words . lines <$> readFile "../inputs/2020/Day2/input.txt"
  -- Part 1
  print $ length . filter id . map (validate . parsePassword) $ content
  -- Part 2
  print $ length . filter id . map (validate2 . parsePassword) $ content

File: typescript/2024/day2.ts
import * as R from 'ramda';

const content = (await Bun.file('../inputs/2024/Day2/input.txt').text()).trim();

const lines = content.split('\n').map(line => line.split(' ').map(x => Number.parseInt(x, 10)));

// console.log(lines);

// part 1

const areAllIncreasing = (ref: number[]) =>
  R.init(ref).reduce((acc, val, i) => {
    if (!acc) return acc;
    const otherVal = ref[i + 1];
    return otherVal >= val + 1 && otherVal <= val + 3;
  }, true);

const areAllDecreasing = (ref: number[]) =>
  R.init(ref).reduce((acc, val, i) => {
    if (!acc) return acc;
    const otherVal = ref[i + 1];
    return otherVal <= val - 1 && otherVal >= val - 3;
  }, true);

const validateRow = (ref: number[]) => areAllIncreasing(ref) || areAllDecreasing(ref);

const r1 = lines.map(validateRow).filter(R.identity<boolean>).length;
console.log(r1);

// part 2

const tryAgainRemovingOneAtATime = (ref: number[]) =>
  R.range(0, ref.length).reduce((acc, idx) => {
    if (acc) return acc;
    const adjusted = R.remove(idx, 1, ref);
    return areAllIncreasing(adjusted) || areAllDecreasing(adjusted);
  }, false);

const validateRow2 = (ref: number[]) =>
  areAllIncreasing(ref) || areAllDecreasing(ref) || tryAgainRemovingOneAtATime(ref);

const r2 = lines.map(validateRow2).filter(R.identity<boolean>).length;
console.log(r2);
